services/galley/src/Galley/API/Public.hs

```diff
@@ -72,6 +87,65 @@ import qualified Wire.API.Team.SearchVisibility as Public
 import qualified Wire.API.User as Public (UserIdList, modelUserIdList)
 import Wire.Swagger (int32Between)
 
+-- This type exists for the special 'HasSwagger' and 'HasServer' instances. It
+-- shows the "Authorization" header in the swagger docs, but expects the
+-- "Z-Auth" header in the server. This helps keep the swagger docs usable
+-- through nginz.
+data ZAuthServant
+
+type InternalAuth = Header' '[Servant.Required, Servant.Strict] "Z-User" UserId
+
+instance HasSwagger api => HasSwagger (ZAuthServant :> api) where
+  toSwagger _ =
+    toSwagger (Proxy @api)
+      & securityDefinitions <>~ InsOrdHashMap.singleton "ZAuth" secScheme
+      & security <>~ [SecurityRequirement $ InsOrdHashMap.singleton "ZAuth" []]
+    where
+      secScheme =
+        SecurityScheme
+          { _securitySchemeType = SecuritySchemeApiKey (ApiKeyParams "Authorization" ApiKeyHeader),
+            _securitySchemeDescription = Just "Must be a token retrieved by calling 'POST /login' or 'POST /access'. It must be presented in this format: 'Bearer \\<token\\>'."
+          }
+
+instance
+  ( HasContextEntry (ctx .++ DefaultErrorFormatters) ErrorFormatters,
+    HasServer api ctx
+  ) =>
+  HasServer (ZAuthServant :> api) ctx
+  where
+  type ServerT (ZAuthServant :> api) m = ServerT (InternalAuth :> api) m
+
+  route _ = Servant.route (Proxy @(InternalAuth :> api))
+  hoistServerWithContext _ pc nt s =
+    Servant.hoistServerWithContext (Proxy @(InternalAuth :> api)) pc nt s
+
+type GetTeamConversationRoles =
+  Summary "Get existing roles available for the given team"
+    :> ZAuthServant
+    :> "teams"
+    :> Capture "tid" TeamId
+    :> "conversations"
+    :> "roles"
+    :> Get '[Servant.JSON] Public.ConversationRolesList
```

Comment:
I think we should use the generics based way of doing these endpoints, similar to what we're doing here: https://github.com/wireapp/wire-server/blob/73b47fce0d53a56b81cf17b7c4e60d0f9f0c80ed/libs/wire-api-federation/src/Wire/API/Federation/API/Brig.hs#L30-L39

This will generate better errors and help us keep all of these `type`s together. I started doing it like this in brig, because I wasn't aware of the other pattern.

Reply:
Agree. Should the apis live in a separate package? Currently you'd need on `brig` for the brig api and on `galley` for the api. If we want to use `servant-client` this could create cyclic dependencies.

Reply:
We could put them into galley-types and brig-types, respectively?

Reply:
What if the APIs lived in `wire-api`?